src/js/heuristicblocking.js

```diff
@@ -310,23 +327,78 @@ HeuristicBlocker.prototype = {
       return; // We already know about the presence of this tracker on the given domain
     }
 
-    // record that we've seen this tracker on this domain (in snitch map)
-    firstParties.push(page_origin);
-    snitchMap.setItem(tracker_origin, firstParties);
+    // If community learning is enabled, queue up a request to the EFF server
+    if (badger.isCommunityLearningEnabled(tab_id)) {
+      let page_fqdn = (new URI(this.tabUrls[tab_id])).host;
+      this.shareTrackerInfo(page_fqdn, tracker_fqdn, tracker_type);
+    }
 
-    // ALLOW indicates this is a tracker still below TRACKING_THRESHOLD
-    // (vs. NO_TRACKING for resources we haven't seen perform tracking yet).
-    // see https://github.com/EFForg/privacybadger/pull/1145#discussion_r96676710
-    this.storage.setupHeuristicAction(tracker_fqdn, constants.ALLOW);
-    this.storage.setupHeuristicAction(tracker_origin, constants.ALLOW);
+    // If local learning is enabled, record that we've seen this tracker on this
+    // domain (in snitch map)
+    if (badger.isLocalLearningEnabled(tab_id)) {
+      firstParties.push(page_origin);
+      snitchMap.setItem(tracker_origin, firstParties);
+
+      // ALLOW indicates this is a tracker still below TRACKING_THRESHOLD
+      // (vs. NO_TRACKING for resources we haven't seen perform tracking yet).
+      // see https://github.com/EFForg/privacybadger/pull/1145#discussion_r96676710
+      this.storage.setupHeuristicAction(tracker_fqdn, constants.ALLOW);
+      this.storage.setupHeuristicAction(tracker_origin, constants.ALLOW);
+
+      // block the origin if it has been seen on multiple first party domains
+      if (firstParties.length >= constants.TRACKING_THRESHOLD) {
+        log('blocklisting origin', tracker_fqdn);
+        this.blocklistOrigin(tracker_origin, tracker_fqdn);
+      }
+    }
+  },
 
-    // Blocking based on outbound cookies
-    var httpRequestPrevalence = firstParties.length;
+  /**
+   * Share information about a tracker for community learning
+   */
+  shareTrackerInfo: function(page_host, tracker_host, tracker_type) {
+    // Share a random sample of trackers we observe
+    if (Math.random() < constants.CL_PROBABILITY) {
+      // check if we've shared this tracker recently
+      // note that this check comes after checking against the snitch map
+      let tr_str = page_host + '+' + tracker_host + '+' + tracker_type;
```

Comment:
Adding some string sanitization and/or sanity checking the input values in this method wouldn't hurt, since they're about to get launched off in a POST